### ISSUE TITLE
False positive finding for `WRONG_INDENTATION`

### ISSUE THREAD

**Issue by WebTiger89** [saveourtool/diktat#1312]
When running diktatCheck I get a false-positive finding for the following code snippet:
```kt
            Surface(
                modifier = Modifier.fillMaxSize(),
                color = MaterialTheme.colors.background,
            ) {
                Text("Android")
            }
```
Warning: expected 20 but was 16
So it expects:
```kt
            Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colors.background,
            ) {
                Text("Android")
            }
```
Maybe this is Jetpack Compose specific and you want to adjust the rule in this direction.

**Comment by akuleshov7** [comment]
Thank you for you report! It is very confusing, but it should be working. We will check asap!

**Comment by akuleshov7** [comment]
@WebTiger89 hm, I checked it in our online demo and it works fine...

![image](https://user-images.githubusercontent.com/58667063/170710821-254d8e09-b1a3-4aa6-9c1f-300b3ec737a9.png)

**Comment by akuleshov7** [comment]
Which version are you using? May be code snippet is more complex?

**Comment by WebTiger89** [comment]
Damn I checked the diktat-analysis.yml file and by default `extendedIndentOfParameters` is true
```
    # If true: in parameter list when parameters are split by newline they are indented with two indentations instead of one
    extendedIndentOfParameters: true
```
When I disable it, it works at this point like expected :)

**Comment by WebTiger89** [comment]
You might want to adjust the default setting. Can be closed. Thanks for the fast support :)

**Comment by akuleshov7** [comment]
> `extendedIndentOfParameters`

yeah, that is the legacy support for old IDEAs :)

**Comment by akuleshov7** [comment]
> You might want to adjust the default setting. 
Exactly! Thank you for catching that! We forgot to revert an example

**Comment by WebTiger89** [comment]
Sorry I have to reopen this. While example above is not reported anymore after setting `extendedIndentOfParameters: false` in config file. Everything else is still reported. When I opened the issue I checked only a small part of the code but now I'm running the check over hundred of files so I noticed it only now.

Example:
```
fun YoutubeChip(selected: Boolean, text: String, modifier: Modifier = Modifier) {
    Surface(
        color = when {
            selected -> MaterialTheme.colors.onSurface
            else -> Color.Transparent
        },
    )
}
```

> [WRONG_INDENTATION] only spaces are allowed for indentation and each indentation should equal to 4 spaces (tabs are not allowed): expected 12 but was 8

etc.

Ktlint online reports it too.

Here the whole rule of my config file:
```
# Checks that indentation is correct
- name: WRONG_INDENTATION
  enabled: true
  configuration:
    # Is newline at the end of a file needed
    newlineAtEnd: true
    # If true: in parameter list when parameters are split by newline they are indented with two indentations instead of one
    extendedIndentOfParameters: false
    # If true: if first parameter in parameter list is on the same line as opening parenthesis, then other parameters can be aligned with it
    alignedParameters: true
    # If true: if expression is split by newline after operator like +/-/`*`, then the next line is indented with two indentations instead of one
    extendedIndentAfterOperators: true
    # The indentation size for each file
    indentationSize: 4
```

**Comment by WebTiger89** [comment]
I might figured out on which code snippets it fails mainly. It seems 2 specific cases can not be handled properly:
1. Expression bodies
2. Jetpack Compose Modifier moved to next line

Example for 1:
```
 val currentTime: Time
    get() =
            with(currentDateTime) {
                Time(hour = hour, minute = minute, second = second)
            }
```
or
```
fun formatDateByPattern(date: String, pattern: String = "ddMMyy"): String =
    DateTimeFormatter.ofPattern(pattern).format(LocalDate.parse(date))
```
or
```
fun LocalDateTime.updateTime(
    hour: Int? = null,
    minute: Int? = null,
    second: Int? = null,
): LocalDateTime = withHour(hour ?: getHour())
    .withMinute(minute ?: getMinute())
    .withSecond(second ?: getSecond())
```
Example for 2:
```
Box(
	Modifier
		.align(Alignment.CenterVertically)
		.widthIn(min = 16.dp + 40.dp),
	contentAlignment = Alignment.CenterStart,
)
```

**Comment by 0x6675636b796f75676974687562** [comment]
@WebTiger89, thank you for your feedback.

I'm looking into the issue and will get back to you in a short while.

**Comment by akuleshov7** [comment]
> fun formatDateByPattern(date: String, pattern: String = "ddMMyy"): String =
>    DateTimeFormatter.ofPattern(pattern).format(LocalDate.parse(date)) 

code provided by @WebTiger89 is a legacy issue in IDEA - I have described it here:
https://github.com/saveourtool/diktat/issues/1323#issuecomment-1141146584
They have partially fixed it and we need to support this logic

**Comment by 0x6675636b796f75676974687562** [comment]
@WebTiger89, today we've made a minor fix (0706885e1eeb0b185853d0e2d0f2bf4fc73238bd) which may affect your experience. From now on, `extendedIndentOfParameters` defaults to `false` even if there's no config file available.

Apart from that, your first example is correct:

```kotlin
@Composable
fun YoutubeChip(selected: Boolean, text: String, modifier: Modifier = Modifier) {
    Surface(
        color = when {
            selected -> MaterialTheme.colors.onSurface
            else -> Color.Transparent
        },
    )
}
```

&mdash; and consistent with the indentation config you've provided:

```yaml
- name: WRONG_INDENTATION
  enabled: true
  configuration:
    newlineAtEnd: true
    extendedIndentOfParameters: false
    alignedParameters: true
    extendedIndentAfterOperators: true
    indentationSize: 4
```

If you change the value of `extendedIndentOfParameters` to `true` (the previous behaviour), the code would need to be reformatted to become

```kotlin
@Composable
fun YoutubeChip(selected: Boolean, text: String, modifier: Modifier = Modifier) {
    Surface(
            color = when {
                selected -> MaterialTheme.colors.onSurface
                else -> Color.Transparent
            },
    )
}
```

&mdash; note that I've increased the indentation in lines 4 through 7: the 4-line `color` expressions is a function parameter and needs to be indented.

Additional examples you've supplied are affected by other flags in the `WRONG_INDENTATION`:

* `alignedParameters`,
* `extendedIndentAfterOperators`, and
* `extendedIndentBeforeDot`.

I'll provide the detailed feedback for each example tomorrow morning. Yet, the way DiKTat's indentation checker behaves looks good to me, I don't see any issues so far.

**Comment by 0x6675636b796f75676974687562** [comment]
Expression bodies, example 1 of 6:

```kotlin
val currentTime: Time
    get() =
            with(currentDateTime) {
                Time(hour = hour, minute = minute, second = second)
            }
```

&mdash; unaffected by the `extendedIndentOfParameters` flag (as well as `alignedParameters` and `extendedIndentBeforeDot`).

If you want to un-set the `extendedIndentAfterOperators` (which is **on** by default), the code needs to be reformatted as follows in order to comply:

```kotlin
val currentTime: Time
    get() =
        with(currentDateTime) {
            Time(hour = hour, minute = minute, second = second)
        }
```

**Comment by 0x6675636b796f75676974687562** [comment]
Expression bodies, example 2 of 6:

```kotlin
fun formatDateByPattern(date: String, pattern: String = "ddMMyy"): String =
    DateTimeFormatter.ofPattern(pattern).format(LocalDate.parse(date))
```

&mdash; same as the previous example, only affected by the `extendedIndentAfterOperators` (needs to be _un-set_ for the formatting to be compliant). If you _set_ the flag or use its default value, be sure to reformat the fragment:

```kotlin
fun formatDateByPattern(date: String, pattern: String = "ddMMyy"): String =
        DateTimeFormatter.ofPattern(pattern).format(LocalDate.parse(date))
```

**Comment by 0x6675636b796f75676974687562** [comment]
Expression bodies, example 3 of 6:

```kotlin
fun LocalDateTime.updateTime(
    hour: Int? = null,
    minute: Int? = null,
    second: Int? = null,
): LocalDateTime = withHour(hour ?: getHour())
    .withMinute(minute ?: getMinute())
    .withSecond(second ?: getSecond())
```

&mdash; with the current formatting **compliant** with `extendedIndentOfParameters` set to **off**, but **not compliant** with the `extendedIndentBeforeDot` flag. If you set `extendedIndentBeforeDot` to **on** (the default), the formatting should be:

```kotlin
fun LocalDateTime.updateTime(
    hour: Int? = null,
    minute: Int? = null,
    second: Int? = null,
): LocalDateTime = withHour(hour ?: getHour())
        .withMinute(minute ?: getMinute())
                .withSecond(second ?: getSecond())
```

With `extendedIndentBeforeDot` set to **off**:

```kotlin
fun LocalDateTime.updateTime(
    hour: Int? = null,
    minute: Int? = null,
    second: Int? = null,
): LocalDateTime = withHour(hour ?: getHour())
    .withMinute(minute ?: getMinute())
        .withSecond(second ?: getSecond())
```

**Comment by 0x6675636b796f75676974687562** [comment]
Expression bodies, example 4 of 6:

![example4-original](https://user-images.githubusercontent.com/73111822/171173102-e5beab20-aa9e-4ea2-b8b2-7ea8f79aa983.svg)

DiKTat reports `WRONG_INDENTATION` errors in lines 17-21 regardless of the `extendedIndentAfterOperators` flag because you're wrapping the expression **before** the `or` operator (or, more precisely, infix function), while wrapping it **after** the operator is the recommended approach.

Wrapping expressions before an operator in Kotlin is discouraged, partially because the same operators may be used for both _unary_ and _binary_ operations, e. g.:

```kotlin
    fun foo() {
        val a = 3 +
                2   // a = 5

        val b = 3   // b = 3
                + 2 // and this is just an unary plus: 2.unaryPlus()
    }
```

For the same reason, DiKTat will additionally report `WRONG_NEWLINES` errors in the same lines (17-21) if this inspection is enabled. I suggest that you reformat your code as follows:

```kotlin
    private fun createLayoutParams(): WindowManager.LayoutParams =
        WindowManager.LayoutParams().apply {
            type = WindowManager.LayoutParams.TYPE_APPLICATION_PANEL
            token = composeView.applicationWindowToken
            width = WindowManager.LayoutParams.MATCH_PARENT
            height = WindowManager.LayoutParams.MATCH_PARENT
            format = PixelFormat.TRANSLUCENT

            // TODO make composable configurable

            // see https://stackoverflow.com/questions/43511326/android-making-activity-full-screen-with-status-bar-on-top-of-it
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                windowInsetsController?.hide(WindowInsets.Type.statusBars())
            } else {
                @Suppress("DEPRECATION")
                systemUiVisibility = (View.SYSTEM_UI_FLAG_LOW_PROFILE or
                        View.SYSTEM_UI_FLAG_FULLSCREEN or
                        View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                        View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or
                        View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION or
                        View.SYSTEM_UI_FLAG_HIDE_NAVIGATION)
            }
        }
```

You will be able to control whether a regular (single) or a _continuation_ (double) indent level is used using the `extendedIndentAfterOperators` flag (**on** by default).

**Comment by 0x6675636b796f75676974687562** [comment]
@WebTiger89, your other examples:

![box](https://user-images.githubusercontent.com/73111822/171395903-c2b4bcba-f2b7-4c3a-b4a7-65397da608da.svg)

and

![surface](https://user-images.githubusercontent.com/73111822/171395957-6b21bbd8-1ada-4ea8-a400-76702c8d3af9.svg)

are well-formatted provided both `extendedIndentOfParameters` and `extendedIndentBeforeDot` are **on**. IDEA code style counterparts are `CONTINUATION_INDENT_IN_ARGUMENT_LISTS` and `CONTINUATION_INDENT_FOR_CHAINED_CALLS`, respectively. As far as I see, DiKTat (both 1.1.0 and 1.1.1-SNAPSHOT) reports no errors against these code fragments.

**Comment by 0x6675636b796f75676974687562** [comment]
The [previous comment](https://github.com/saveourtool/diktat/issues/1312#issuecomment-1143643806) turned out to be not true. IDEA behaves **differently**, depending on _how exactly_ Kotlin style guide is applied.

By default (i. e. unless *any* style guide is applied, empty XML file), _continuation indent_ is **on** wherever applicable, equivalent to:

```xml
    <JetCodeStyleSettings>
      <option name="CONTINUATION_INDENT_IN_PARAMETER_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_IN_ARGUMENT_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_FOR_EXPRESSION_BODIES" value="true" />
      <option name="CONTINUATION_INDENT_FOR_CHAINED_CALLS" value="true" />
      <option name="CONTINUATION_INDENT_IN_SUPERTYPE_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_IN_IF_CONDITIONS" value="true" />
      <option name="CONTINUATION_INDENT_IN_ELVIS" value="true" />
    </JetCodeStyleSettings>
```

Now, if the code style is set via _Preferences -> Editor -> Code Style -> Kotlin -> Load/Save -> Use defaults from: -> Kotlin Coding Conventions_:

![idea-code-style-0](https://user-images.githubusercontent.com/73111822/171611579-9ad65cfe-3786-47d0-9aca-6a1e97143608.png)

&mdash; then the following settings are explicitly set in XML:

```xml
    <JetCodeStyleSettings>
      <option name="CONTINUATION_INDENT_IN_PARAMETER_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_IN_ARGUMENT_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_FOR_EXPRESSION_BODIES" value="true" />
      <option name="CONTINUATION_INDENT_FOR_CHAINED_CALLS" value="true" />
      <option name="CONTINUATION_INDENT_IN_SUPERTYPE_LISTS" value="true" />
      <option name="CONTINUATION_INDENT_IN_IF_CONDITIONS" value="true" />
      <option name="CONTINUATION_INDENT_IN_ELVIS" value="true" />
      <option name="CODE_STYLE_DEFAULTS" value="KOTLIN_OFFICIAL" />
    </JetCodeStyleSettings>
```

Note that `CODE_STYLE_DEFAULTS` is set to `KOTLIN_OFFICIAL`, and all `CONTINUATION_INDENT_*` flags are *explicitly* set to `true`. 

Next, if we re-set our code style settings to the defaults and re-apply the same Kotlin style guide, now from the "Set from..." hyperlink, then `.idea/codeStyles/Project.xml` will have the following content:

```xml
    <JetCodeStyleSettings>
      <option name="CODE_STYLE_DEFAULTS" value="KOTLIN_OFFICIAL" />
    </JetCodeStyleSettings>
```

&mdash; and the effective values of all _continuation indent_ related flags will be `false`. 

That said, IDEA behaviour is inconsistent and misleading, and the [official style guide](https://kotlinlang.org/docs/coding-conventions.html) (in which there's not a single mention of a _continuation indent_) should be consulted from now on.
